# Municipal waste generated by countries

### Content
<ol>
    <li>Introduction</li>
    <li>Data description and objectives</li>
    <li>Data manipulation and validation</li>
    <ol>
        <li> Data cleaning & shaping </li>
        <li>Missing values</li>
    </ol>
<li>Web-Scrapping for missing values</li>
</ol>

## 1.Introduction

The world generates millions of tonnes of solid waste every year. Out daily life consist of many things, that we eat, break, use, and then drop it out, because these items already are unnecessary and unuseful. This way humanity creates mountains of waste right behind their living places. 

Municipal waste management in the world has become more and more complex in the last decade because annually, the number of municipal solid waste is rapidly growing, but a tiny portion of it is recycled by countries. So, the aim of this project is to provide guidelines on the scope and coverage of municipal waste for municipal waste collection. 

This investigation is based on statistical data, which in the project will be focusing on how much waste generates humanity.

(https://en.wikipedia.org/wiki/Municipal_solid_waste)

## 2. Data description and objectives

##### Description

Dataset, which was chosen for investigation of the topic, is about country level municipal waste, that is generated by particular country, was found on the Word Bank web-site (https://datacatalog.worldbank.org/dataset/what-waste-global-database/resource/de74cc68-e796-4e42-9793-f140719c91ac#{}). There are 271 rows, and 51 columns in total. However, for this project only necessary and meaningful columns are taken. 
Out of these 51 columns, were chosen 11 columns. 
They are: 
* Country - name of country
* GDP ($M) - GDP of this country
* Food_Organic (%) - percentage of food and organic waste
* Glass(%) - percentage of glass waste
* Metal (%) - percentage of metal waste
* Other(%) - percentage of waste, which is not belongs to any large category
* Paper (%) - percentage of paper waste
* Plastic (%) - percentage of  plastic waste
* Population (%) - number of people living in the country
* Total_Waste (tonnes) - mass of all categories of waste in tonnes
* Recycling (%) - percentage of waste that recycled



##### Objectives

1. Observe the dependence of GDP per capita on the amount of waste generated by one person
2. According to the world map, investigate which countries produce the most amount of waste
3. Consider the relationship between the GDP of the country with recycling waste in percentages
4. Investigate which category of waste (Food/Organic, Glass, Metal, Other, Paper, Plastic) is in the lead 
5. Investigate the factors that affect the production of municipal waste in countries

## 3. Data manipulation and validation

In [1]:
#import libraries
import pandas as pd
import requests
from matplotlib import pyplot as plt
import numpy as np
from bs4 import BeautifulSoup

In [2]:
#reading a csv file
waste = pd.read_csv('country_level_data_0.csv')
waste

iso3c region_id  country_name income_id           gdp  \
0     ABW       LCN         Aruba       HIC           NaN   
1     AFG       SAS   Afghanistan       LIC  2.141361e+10   
2     AGO       SSF        Angola       LMC  1.030423e+11   
3     ALB       ECS       Albania       UMC  1.347108e+10   
4     AND       ECS       Andorra       HIC  3.319880e+09   
..    ...       ...           ...       ...           ...   
212   XKX       ECS        Kosovo       LMC  7.129272e+09   
213   YEM       MEA   Yemen, Rep.       LIC  1.192703e+10   
214   ZAF       SSF  South Africa       UMC  4.212087e+11   
215   ZMB       SSF        Zambia       LMC  2.703717e+10   
216   ZWE       SSF      Zimbabwe       LIC  1.481899e+10   

     composition_food_organic_waste_percent  composition_glass_percent  \
0                                       NaN                        NaN   
1                                       NaN                        NaN   
2                                 51.800000                   6.700000   
3                                 51.400000                   4.500000   
4                                 31.200000                   8.200000   
..                                      ...                        ...   
212                               42.000000                   6.000000   
213                               65.000000                   1.000000   
214                               16.381655                   5.200216   
215                                     NaN                        NaN   
216                               36.000000                   5.000000   

     composition_metal_percent  composition_other_percent  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                     4.400000                  11.500000   
3                     4.800000                  15.210000   
4                     2.600000                  11.600000   
..                         ...                        ...   
212                   6.000000                  20.000000   
213                   6.000000                   6.000000   
214                  16.910461                  45.020646   
215                        NaN                        NaN   
216                   6.000000                   3.000000   

     composition_paper_cardboard_percent  ...  \
0                                    NaN  ...   
1                                    NaN  ...   
2                              11.900000  ...   
3                               9.900000  ...   
4                              35.100000  ...   
..                                   ...  ...   
212                             8.000000  ...   
213                             7.000000  ...   
214                             9.396918  ...   
215                                  NaN  ...   
216                            27.000000  ...   

     waste_treatment_controlled_landfill_percent  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   
..                                           ...   
212                                        66.43   
213                                        12.00   
214                                        72.00   
215                                          NaN   
216                                          NaN   

     waste_treatment_incineration_percent  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                    52.1   
..                                    ...   
212                                   NaN   
213                                   NaN   
214                           

In [3]:
#shape of the file
waste.shape

(217, 51)

### 3.1 Data Cleaning and Shaping

#### Deleting unnecessary columns in the "waste" dataset

In [4]:
waste = waste.drop(['iso3c', 'income_id','waste_treatment_controlled_landfill_percent',
                   'waste_treatment_incineration_percent',
                   'waste_treatment_landfill_unspecified_percent',
                   'waste_treatment_open_dump_percent',
                   'waste_treatment_other_percent',
                   'waste_treatment_sanitary_landfill_landfill_gas_system_percent',
                   'waste_treatment_unaccounted_for_percent'], 1)

In [5]:
waste = waste.drop(['waste_collection_coverage_total_percent_of_waste', 
                   'waste_collection_coverage_urban_percent_of_households',
                   'waste_collection_coverage_urban_percent_of_geographic_area',
                   'waste_collection_coverage_urban_percent_of_population',
                   'waste_collection_coverage_urban_percent_of_waste',
                   'waste_treatment_anaerobic_digestion_percent',
                   'waste_treatment_compost_percent',
                   'waste_treatment_waterways_marine_percent',
                   'where_where_is_this_data_measured'], 1)



In [6]:
waste = waste.drop(['waste_collection_coverage_rural_percent_of_geographic_area', 
                   'waste_collection_coverage_rural_percent_of_households',
                   'waste_collection_coverage_rural_percent_of_population',
                   'waste_collection_coverage_rural_percent_of_waste',
                   'waste_collection_coverage_total_percent_of_geographic_area',
                   'waste_collection_coverage_total_percent_of_households',
                   'waste_collection_coverage_total_percent_of_population'], 1)

In [7]:
waste = waste.drop(['special_waste_hazardous_waste_tons_year', 
                   'special_waste_e_waste_tons_year',
                   'special_waste_construction_and_demolition_waste_tons_year',
                   'other_information_summary_of_key_solid_waste_information_made_available_to_the_public'], 1)



In [8]:
waste = waste.drop(['region_id', 
                   'other_information_information_system_for_solid_waste_management',
                   'other_information_national_agency_to_enforce_solid_waste_laws_and_regulations',
                   'other_information_national_law_governing_solid_waste_management_in_the_country',
                   'other_information_ppp_rules_and_regulations'], 1)

In [9]:
waste = waste.drop(['special_waste_agricultural_waste_tons_year', 
                   'special_waste_industrial_waste_tons_year',
                   'special_waste_medical_waste_tons_year'], 1)

In [10]:
waste = waste.drop(['composition_rubber_leather_percent', 
                   'composition_wood_percent',
                   'composition_yard_garden_green_waste_percent'], 1)

In [11]:
#showing a new shape of the table
waste.shape

(217, 11)

#### Rename colums in appropriate way with measurements in brackets

In [12]:
waste.columns = ["Country", "GDP ($M)", "Food_Organic (%)", "Glass (%)", "Metal (%)" , "Other (%)", "Paper (%)", "Plastic (%)",
                "Population (millions)", "Total Waste (tonnes)", "Recycling (%)"]
waste

Country      GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0           Aruba           NaN               NaN        NaN        NaN   
1     Afghanistan  2.141361e+10               NaN        NaN        NaN   
2          Angola  1.030423e+11         51.800000   6.700000   4.400000   
3         Albania  1.347108e+10         51.400000   4.500000   4.800000   
4         Andorra  3.319880e+09         31.200000   8.200000   2.600000   
..            ...           ...               ...        ...        ...   
212        Kosovo  7.129272e+09         42.000000   6.000000   6.000000   
213   Yemen, Rep.  1.192703e+10         65.000000   1.000000   6.000000   
214  South Africa  4.212087e+11         16.381655   5.200216  16.910461   
215        Zambia  2.703717e+10               NaN        NaN        NaN   
216      Zimbabwe  1.481899e+10         36.000000   5.000000   6.000000   

     Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0          NaN        NaN          NaN              103187.00   
1          NaN        NaN          NaN            34656032.00   
2    11.500000  11.900000    13.500000            25096150.00   
3    15.210000   9.900000     9.600000             2880703.00   
4    11.600000  35.100000    11.300000               82431.00   
..         ...        ...          ...                    ...   
212  20.000000   8.000000    11.000000             1801800.00   
213   6.000000   7.000000    10.000000            27584213.00   
214  45.020646   9.396918     7.090104            51729345.36   
215        NaN        NaN          NaN            14264756.00   
216   3.000000  27.000000    23.000000            12500525.00   

     Total Waste (tonnes)  Recycling (%)  
0            8.813202e+04           11.0  
1            5.628525e+06            NaN  
2            4.213644e+06            NaN  
3            1.142964e+06            NaN  
4            4.300000e+04            NaN  
..                    ...            ...  
212          3.190000e+05            NaN  
213          4.836820e+06            8.0  
214          1.845723e+07           28.0  
215          2.608268e+06            NaN  
216          1.449752e+06           16.0  

[217 rows x 11 columns]

#### Convernt all values in a table from Scentific Notation to float

In [13]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)
waste

Country        GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0           Aruba             nan               nan        nan        nan   
1     Afghanistan  21413614653.32               nan        nan        nan   
2          Angola 103042328743.66             51.80       6.70       4.40   
3         Albania  13471082475.18             51.40       4.50       4.80   
4         Andorra   3319880351.13             31.20       8.20       2.60   
..            ...             ...               ...        ...        ...   
212        Kosovo   7129271791.57             42.00       6.00       6.00   
213   Yemen, Rep.  11927030389.45             65.00       1.00       6.00   
214  South Africa 421208662669.69             16.38       5.20      16.91   
215        Zambia  27037168289.17               nan        nan        nan   
216      Zimbabwe  14818986230.97             36.00       5.00       6.00   

     Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0          nan        nan          nan              103187.00   
1          nan        nan          nan            34656032.00   
2        11.50      11.90        13.50            25096150.00   
3        15.21       9.90         9.60             2880703.00   
4        11.60      35.10        11.30               82431.00   
..         ...        ...          ...                    ...   
212      20.00       8.00        11.00             1801800.00   
213       6.00       7.00        10.00            27584213.00   
214      45.02       9.40         7.09            51729345.36   
215        nan        nan          nan            14264756.00   
216       3.00      27.00        23.00            12500525.00   

     Total Waste (tonnes)  Recycling (%)  
0                88132.02          11.00  
1              5628525.37            nan  
2              4213643.58            nan  
3              1142964.00            nan  
4                43000.00            nan  
..                    ...            ...  
212             319000.00            nan  
213            4836820.00           8.00  
214           18457232.00          28.00  
215            2608268.00            nan  
216            1449752.00          16.00  

[217 rows x 11 columns]

#### Convert GDP to Millions of Dollars

In [14]:
waste['GDP ($M)'] = waste['GDP ($M)'].div(1000000000)
waste

Country  GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0           Aruba       nan               nan        nan        nan   
1     Afghanistan     21.41               nan        nan        nan   
2          Angola    103.04             51.80       6.70       4.40   
3         Albania     13.47             51.40       4.50       4.80   
4         Andorra      3.32             31.20       8.20       2.60   
..            ...       ...               ...        ...        ...   
212        Kosovo      7.13             42.00       6.00       6.00   
213   Yemen, Rep.     11.93             65.00       1.00       6.00   
214  South Africa    421.21             16.38       5.20      16.91   
215        Zambia     27.04               nan        nan        nan   
216      Zimbabwe     14.82             36.00       5.00       6.00   

     Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0          nan        nan          nan              103187.00   
1          nan        nan          nan            34656032.00   
2        11.50      11.90        13.50            25096150.00   
3        15.21       9.90         9.60             2880703.00   
4        11.60      35.10        11.30               82431.00   
..         ...        ...          ...                    ...   
212      20.00       8.00        11.00             1801800.00   
213       6.00       7.00        10.00            27584213.00   
214      45.02       9.40         7.09            51729345.36   
215        nan        nan          nan            14264756.00   
216       3.00      27.00        23.00            12500525.00   

     Total Waste (tonnes)  Recycling (%)  
0                88132.02          11.00  
1              5628525.37            nan  
2              4213643.58            nan  
3              1142964.00            nan  
4                43000.00            nan  
..                    ...            ...  
212             319000.00            nan  
213            4836820.00           8.00  
214           18457232.00          28.00  
215            2608268.00            nan  
216            1449752.00          16.00  

[217 rows x 11 columns]

#### Covert column of population to millions of people

In [15]:
waste['Population (millions)'] = waste['Population (millions)'].div(1000000)

In [16]:
#round the number in order to get integer number with two zero(.00) decimal places
waste['Total Waste (tonnes)'] =  waste['Total Waste (tonnes)'].round(1)
waste

Country  GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0           Aruba       nan               nan        nan        nan   
1     Afghanistan     21.41               nan        nan        nan   
2          Angola    103.04             51.80       6.70       4.40   
3         Albania     13.47             51.40       4.50       4.80   
4         Andorra      3.32             31.20       8.20       2.60   
..            ...       ...               ...        ...        ...   
212        Kosovo      7.13             42.00       6.00       6.00   
213   Yemen, Rep.     11.93             65.00       1.00       6.00   
214  South Africa    421.21             16.38       5.20      16.91   
215        Zambia     27.04               nan        nan        nan   
216      Zimbabwe     14.82             36.00       5.00       6.00   

     Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0          nan        nan          nan                   0.10   
1          nan        nan          nan                  34.66   
2        11.50      11.90        13.50                  25.10   
3        15.21       9.90         9.60                   2.88   
4        11.60      35.10        11.30                   0.08   
..         ...        ...          ...                    ...   
212      20.00       8.00        11.00                   1.80   
213       6.00       7.00        10.00                  27.58   
214      45.02       9.40         7.09                  51.73   
215        nan        nan          nan                  14.26   
216       3.00      27.00        23.00                  12.50   

     Total Waste (tonnes)  Recycling (%)  
0                88132.00          11.00  
1              5628525.40            nan  
2              4213643.60            nan  
3              1142964.00            nan  
4                43000.00            nan  
..                    ...            ...  
212             319000.00            nan  
213            4836820.00           8.00  
214           18457232.00          28.00  
215            2608268.00            nan  
216            1449752.00          16.00  

[217 rows x 11 columns]

### 3.2 Missing values

#### Drop rows of NaN

In [17]:
nan_df = waste[waste.isna().any(axis=1)]

waste = waste.dropna(how='all')
waste.head(70)

Country  GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0                   Aruba       nan               nan        nan        nan   
1             Afghanistan     21.41               nan        nan        nan   
2                  Angola    103.04             51.80       6.70       4.40   
3                 Albania     13.47             51.40       4.50       4.80   
4                 Andorra      3.32             31.20       8.20       2.60   
..                    ...       ...               ...        ...        ...   
65         Faeroe Islands       nan               nan        nan        nan   
66  Micronesia, Fed. Sts.      0.30             23.83       7.07      16.73   
67                  Gabon     18.91               nan        nan        nan   
68         United Kingdom   2757.62             16.70       2.20       3.50   
69                Georgia     15.19             39.00       3.00       5.00   

    Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0         nan        nan          nan                   0.10   
1         nan        nan          nan                  34.66   
2       11.50      11.90        13.50                  25.10   
3       15.21       9.90         9.60                   2.88   
4       11.60      35.10        11.30                   0.08   
..        ...        ...          ...                    ...   
65        nan        nan          nan                   0.05   
66       9.32      13.30        26.17                   0.10   
67        nan        nan          nan                   1.09   
68      28.20      18.90        20.20                  65.13   
69      16.00      34.00         3.00                   3.72   

    Total Waste (tonnes)  Recycling (%)  
0               88132.00          11.00  
1             5628525.40            nan  
2             4213643.60            nan  
3             1142964.00            nan  
4               43000.00            nan  
..                   ...            ...  
65              61000.00          67.00  
66              26039.60            nan  
67             238102.30            nan  
68           31567000.00          27.25  
69             800000.00            nan  

[70 rows x 11 columns]

#### Drop all records, where the percentage of waste in each category is NaN, and where country is a NaN value

In [18]:
waste = waste[waste['Food_Organic (%)'].notna()
              & waste['Glass (%)'].notna()
              & waste['Metal (%)'].notna()
              & waste['Other (%)'].notna()
              & waste['Paper (%)'].notna()
              & waste['Plastic (%)'].notna()]
waste = waste[waste['Country'].notna()]
        

#### Reset indeces because of deleteming missing values in the above example

In [19]:
waste = waste.reset_index(drop = True)
waste

Country  GDP ($M)  Food_Organic (%)  Glass (%)  Metal (%)  \
0                  Angola    103.04             51.80       6.70       4.40   
1                 Albania     13.47             51.40       4.50       4.80   
2                 Andorra      3.32             31.20       8.20       2.60   
3    United Arab Emirates    384.22             39.00       4.00       3.00   
4               Argentina    447.52             38.74       3.16       1.84   
..                    ...       ...               ...        ...        ...   
155                 Samoa      0.74             42.60       2.20       8.80   
156                Kosovo      7.13             42.00       6.00       6.00   
157           Yemen, Rep.     11.93             65.00       1.00       6.00   
158          South Africa    421.21             16.38       5.20      16.91   
159              Zimbabwe     14.82             36.00       5.00       6.00   

     Other (%)  Paper (%)  Plastic (%)  Population (millions)  \
0        11.50      11.90        13.50                  25.10   
1        15.21       9.90         9.60                   2.88   
2        11.60      35.10        11.30                   0.08   
3        10.00      25.00        19.00                   9.27   
4        15.36      13.96        14.61                  42.98   
..         ...        ...          ...                    ...   
155      19.40       7.20        13.00                   0.19   
156      20.00       8.00        11.00                   1.80   
157       6.00       7.00        10.00                  27.58   
158      45.02       9.40         7.09                  51.73   
159       3.00      27.00        23.00                  12.50   

     Total Waste (tonnes)  Recycling (%)  
0              4213643.60            nan  
1              1142964.00            nan  
2                43000.00            nan  
3              5413453.40          20.00  
4             17910550.00           6.00  
..                    ...            ...  
155              27399.10          36.00  
156             319000.00            nan  
157            4836820.00           8.00  
158           18457232.00          28.00  
159            1449752.00          16.00  

[160 rows x 11 columns]

#### Finding indexes where GDP of a country is NaN value

In [20]:
# here we try to find all missing values in GDP column 
# in order to get its indeces for the further web-scpaping
column = pd.isna(waste['GDP ($M)'])
gdp_nan = []
for row in range(2, len(waste)):
    if column[row] == True:
        gdp_nan.insert(row, row)
print(gdp_nan)

#the result gives us an array, where country has no value for GDP

[21, 35, 36, 56, 87, 104, 119, 132, 134, 135, 151, 152]


#### Array of all countries in the WASTE tabel

In [21]:
countries = []
for row in range(len(waste)):
    countries.insert(row, waste['Country'][row])
print(countries)                 

['Angola', 'Albania', 'Andorra', 'United Arab Emirates', 'Argentina', 'Armenia', 'American Samoa', 'Antigua and Barbuda', 'Australia', 'Austria', 'Azerbaijan', 'Burundi', 'Belgium', 'Benin', 'Burkina Faso', 'Bangladesh', 'Bulgaria', 'Bahrain', 'Bahamas, The', 'Belarus', 'Belize', 'Bermuda', 'Bolivia', 'Brazil', 'Barbados', 'Brunei Darussalam', 'Bhutan', 'Canada', 'Switzerland', 'Chile', 'China', 'Cameroon', 'Colombia', 'Comoros', 'Costa Rica', 'Cuba', 'Cayman Islands', 'Cyprus', 'Czech Republic', 'Germany', 'Dominica', 'Denmark', 'Dominican Republic', 'Algeria', 'Ecuador', 'Egypt, Arab Rep.', 'Spain', 'Estonia', 'Ethiopia', 'Finland', 'Fiji', 'France', 'Micronesia, Fed. Sts.', 'United Kingdom', 'Georgia', 'Ghana', 'Gibraltar', 'Guinea', 'Greece', 'Grenada', 'Greenland', 'Guam', 'Guyana', 'Hong Kong SAR, China', 'Honduras', 'Croatia', 'Haiti', 'Hungary', 'Indonesia', 'Ireland', 'Iran, Islamic Rep.', 'Iraq', 'Israel', 'Italy', 'Jamaica', 'Jordan', 'Japan', 'Kazakhstan', 'Kenya', 'Cambodi

#### Retrieving and insert in the new array only those countries, where GDP value is NaN

In [22]:
countries_nan = []
for row in range(len(gdp_nan)):
    countries_nan.insert(row, countries[gdp_nan[row]])
print(countries_nan)

# here the result of countries that has no GDP values

['Bermuda', 'Cuba', 'Cayman Islands', 'Gibraltar', 'Liechtenstein', 'New Caledonia', 'French Polynesia', 'Sint Maarten (Dutch part)', 'Syrian Arab Republic', 'Turks and Caicos Islands', 'British Virgin Islands', 'Virgin Islands (U.S.)']


## 4. Web-Scraping for missing values

In this step, I scrape (https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)) web-site, to find GDP of countries that have missing values (countries_nan - array, which is mentioned above). Then I want to join these two tables, if there will be the same countries.

In [49]:
#200 successful connection to this web-page
wikiGDP = requests.get('https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)')
wikiGDP.status_code

200

In [50]:
#find table number 8 on this page (that has necessary data for this project)
soup = BeautifulSoup(wikiGDP.text, 'html.parser')
table = soup.find_all('table')[8]
print(table)

<table class="wikitable sortable" style="margin-left:auto; margin-right:auto; margin-top:0;">
<tbody><tr>
<th data-sort-type="number" style="width:2em;">Rank</th>
<th>Country/Territory</th>
<th>GDP<br/>(US$million)
</th></tr>
<tr style="font-weight:bold;background:#eaecf0">
<td data-sort-value="-1"> </td>
<td data-sort-value=""><span class="flagicon" style="padding-left:25px;"> </span><i><a href="/wiki/Gross_world_product" title="Gross world product">World</a></i></td>
<td align="right"><span data-sort-value="7007877515410000000♠">87,751,541</span>
</td></tr>
<tr>
<td align="right">1</td>
<td><span class="flagicon"><img alt="" class="thumbborder" data-file-height="650" data-file-width="1235" decoding="async" height="12" src="//upload.wikimedia.org/wikipedia/en/thumb/a/a4/Flag_of_the_United_States.svg/23px-Flag_of_the_United_States.svg.png" srcset="//upload.wikimedia.org/wikipedia/en/thumb/a/a4/Flag_of_the_United_States.svg/35px-Flag_of_the_United_States.svg.png 1.5x, //upload.wikimedia

In [77]:
#getting all information from this web-page
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)')[8]
df

Rank        Country/Territory  GDP(US$million)
0    NaN                    World         87751541
1      1            United States         21427700
2      2               China[n 5]         14342903
3      3                    Japan          5081770
4      4                  Germany          3845630
..   ...                      ...              ...
186  181             Palau (2018)              284
187  182  Marshall Islands (2018)              221
188  183                 Kiribati              195
189  184                    Nauru              118
190  185                   Tuvalu               47

[191 rows x 3 columns]

In [78]:
#deleting the first record
df = df.dropna(how='any')

In [79]:
# set Rank as an index column
df = df.set_index('Rank')

In [82]:
df

Country/Territory  GDP(US$million)
Rank                                          
1               United States         21427700
2                  China[n 5]         14342903
3                       Japan          5081770
4                     Germany          3845630
5                       India          2875142
...                       ...              ...
181              Palau (2018)              284
182   Marshall Islands (2018)              221
183                  Kiribati              195
184                     Nauru              118
185                    Tuvalu               47

[190 rows x 2 columns]

As it seen we have 190 row, but the last Rank is 185. That means thasomewhere in the dataframe we have missed Rank value.

In [84]:
#now we reseted the indeces of the dataframe, and have the same number of records as number of ranks
#exacly 190rows and ranks
df = df.reset_index(drop = True)
df

Country/Territory  GDP(US$million)
0              United States         21427700
1                 China[n 5]         14342903
2                      Japan          5081770
3                    Germany          3845630
4                      India          2875142
..                       ...              ...
185             Palau (2018)              284
186  Marshall Islands (2018)              221
187                 Kiribati              195
188                    Nauru              118
189                   Tuvalu               47

[190 rows x 2 columns]

## Working with recycling table

Taking only data with the max year(last)

In [16]:
recycling = recycling.drop(['Code'], 1);

In [17]:
total = 678;
country = 'Australia'
dicts = {}
values = recycling['Entity']
for i in range(total):
    if values[i] != values[i+1]:
        dicts[i] = values[i] 
print(dicts)



{3: 'Australia', 29: 'Austria', 55: 'Belgium', 56: 'Brazil', 65: 'Canada', 66: 'Costa Rica', 87: 'Czech Republic', 110: 'Denmark', 131: 'Estonia', 153: 'Finland', 175: 'France', 197: 'Germany', 217: 'Greece', 240: 'Hungary', 257: 'Iceland', 275: 'Ireland', 277: 'Israel', 298: 'Italy', 323: 'Japan', 346: 'Korea', 361: 'Latvia', 382: 'Lithuania', 403: 'Luxembourg', 425: 'Mexico', 449: 'Netherlands', 474: 'Norway', 495: 'Poland', 515: 'Portugal', 533: 'Slovakia', 554: 'Slovenia', 562: 'Spain', 585: 'Sweden', 611: 'Switzerland', 632: 'Turkey', 653: 'United Kingdom'}


In [28]:
dicts.values()

dict_values(['Australia', 'Austria', 'Belgium', 'Brazil', 'Canada', 'Costa Rica', 'Czech Republic', 'Denmark', 'Estonia', 'Finland', 'France', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland', 'Israel', 'Italy', 'Japan', 'Korea', 'Latvia', 'Lithuania', 'Luxembourg', 'Mexico', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland', 'Turkey', 'United Kingdom'])

In [18]:
dicts.get(3)
print(len(dicts))
for i in range(total):    
    if dicts.get(i) != None:
        print(dicts.get(i))
        

35
Australia
Austria
Belgium
Brazil
Canada
Costa Rica
Czech Republic
Denmark
Estonia
Finland
France
Germany
Greece
Hungary
Iceland
Ireland
Israel
Italy
Japan
Korea
Latvia
Lithuania
Luxembourg
Mexico
Netherlands
Norway
Poland
Portugal
Slovakia
Slovenia
Spain
Sweden
Switzerland
Turkey
United Kingdom


In [54]:
page = requests.get('https://knoema.com/search?query=world%20waste%20municipal%20waste%20collected%20which%20is%20recycled&scope=timeseries')
page.status_code

200

In [73]:
soup = BeautifulSoup(page.text, 'html.parser')

In [74]:
page.text

'\r\n\r\n<!DOCTYPE html>\r\n<html lang="en-US" dir="ltr"  class="unknown unknown0" \txmlns:fb="http://www.facebook.com/2008/fbml" >\r\n<head runat="server" \tprefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb# knoema-com: http://ogp.me/ns/fb/knoema-com#" >\r\n    \t\t<!-- Google Tag Manager -->\n\t\t<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({\'gtm.start\':\n\t\tnew Date().getTime(),event:\'gtm.js\'});var f=d.getElementsByTagName(s)[0],\n\t\tj=d.createElement(s),dl=l!=\'dataLayer\'?\'&l=\'+l:\'\';j.async=true;j.src=\n\t\t\'https://www.googletagmanager.com/gtm.js?id=\'+i+dl;f.parentNode.insertBefore(j,f);\n\t\t})(window,document,\'script\',\'dataLayer\',\'GTM-NTMKTKM\');</script>\n\t\t<!-- End Google Tag Manager -->\n\r\n    \r\n\t\r\n\r\n    <meta charset="utf-8" /><script type="text/javascript">window.NREUM||(NREUM={});NREUM.info = {"beacon":"bam.nr-data.net","errorBeacon":"bam.nr-data.net","licenseKey":"313cca50ec","applicationID":"1572704","transactionName":"NQdUMkBQCkM

In [78]:
row = soup.find_all('tr')
print(row)

[]


In [33]:
bs4 = BeautifulSoup(req_pop.content,'lxml')
txt = '''
<td class="series-column">
    <a class="formatted-title external-link result-url" href="//">link text</a>
</td>'''

soup = BeautifulSoup(txt, 'html.parser')

for link in soup.select('td.series-column a.formatted-title.external-link.result-url'):
    print(link['href'])


In [72]:
bs4 = BeautifulSoup(url)
print(bs4)
bs4.find_all('td', class_='series-column')
    

<html><body><p>https://knoema.com/search?query=world%20waste%20municipal%20waste%20collected%20which%20is%20recycled&amp;scope=timeseries</p></body></html>


[]